### ISSUE TITLE
Improve search/filter functionality

### ISSUE THREAD

**Issue by uruvelakassapa** [bkhpanigha/hh-suttas#252]
I'm wondering if it's possible to merge the functionality of both searches into one bar/screen. Currently I feel the advanced search is somewhat hard to find. I have no notion of how search is implemented so I dont know if advanced is more resource intensive and thus it's not feasible.

But I was thinking if instead of seeing the following:
![Screenshot_20241226_174636_Chrome](https://github.com/user-attachments/assets/f8a72092-97a4-4f47-a0f9-62e801c26f4f)

We could just see this immediately (perhaps one can add a delay to start the search so to give the user some time to type out the whole word. Also the search settings could the a toggle like actual advanced search is at the moment):

![Screenshot_20241226_174700_Chrome](https://github.com/user-attachments/assets/64d44c2e-7a2f-4274-962d-09dd444d32c0)

Additionally we could have a seperate section or mode if the user types a citation which would be quite easy to filter on (starts mn, dn etc)

![Screenshot_20241226_174757_Chrome](https://github.com/user-attachments/assets/c70f128a-61c7-486b-aecb-171d31bb2ae1)

Just a suggestion for new users of the site. Thanks for everyone's contributions on the search functionality. I'm finding it very useful!

**Comment by Reptilioo** [comment]
Yes that could be a good idea if new users would find it handy while not knowing where to find it.
I'm not a big fan of delay before search, especially if we need to go to a new page like it is the case here, but a simple search button could do the trick, provided people that are used to the automatically started search don't think it is broken.

Just for info: Resource-wise, the user need to have downloaded the entire database on their end before being able to use the advanced search, which is something that is done automatically when visiting the website as of now. Meanwhile the user only need to download the `python/generated/available_suttas.json` file to use the standard function.

I believe there was the idea to implement a version for people with limited bandwith using only the standard function, this would most likely not interfer with this project since we would still be able to test if user had "offline mode" activated and use the adequate search from there.

**Comment by uruvelakassapa** [comment]
Thanks for the context. I have a couple of days with a computer so I'll have a look to see how big the search footprint is at the moment given we are maybe at 10% of all the peak Sutta content. If it's small enough then maybe we can ditch the limited bandwidth distinction.

**Comment by uruvelakassapa** [comment]
@Reptilioo So I didn't get around to having a look at the space topic. However, one option around it could be to merge it for now, as we are downloading the whole site anyway, and then let the user decide whether they want the full version or the lite version. This would be much easier to explain once we have the downloads page from #257

**Comment by interested-in-dhamma** [comment]
@Reptilioo @bkhpanigha 
When I tested it now, I think removing that filter functionality is not so great. But I don't know if I was the only one using it.

Also the way it is done now in general is just kind of confusing, entering search (on home page), then when searching being taken to adv search page with a complete different interface. I think it is just too much unnecessary complexity.
In general it is always good to keep the designs simple.

I have some ideas to simplify it if you like to go over it.
I would change it like this:
standard search = filter
advanced search = search

I would make the search part of the header, when clicked it pops out (think like the https://suttacentral.net/?lang=en search), the pop-out is the advanced search page you added. (So no longer have it like a link)

The filter input I would make like the example here https://icon-sets.iconify.design/ (see the filter on the left)
And I would make it filter only on citation and title again as it was in the beginning.

**Comment by bkhpanigha** [comment]
@Reptilioo @interested-in-dhamma 

I agree that it might be good to reintroduce the filter functionality, but just for the titles and not the Sutta contents (and that it can match titles in Pali as well even though they're not displayed).

**Comment by Reptilioo** [comment]
@bkhpanigha @interested-in-dhamma 

Sure, let's do that, I see 3 options right now:

1. Put back the standard filter-like search (for eng/pali titles + headings) but keep the other behaviors, meaning the user can use the filter functionality and will still be sent to the search panel if they press Enter or click search button. 
**[Easier but potentially confusing for the user (?)]**

2. Revert back to the previous version and add a less discreet link to the search panel, probably as a top button (for now) and as a text link in the future redesigned header. 
**[Simple]**

3. Interested-in-dhamma idea: one field above the suttas list used as a filter, and one button in the header opening a pop-up/modal box containing the search panel, from which the user can start a search. 
**[Sophisticated]**


If (2) or (3) is preferred we can revert back to the previous version and go from there.

**Comment by bkhpanigha** [comment]
@Reptilioo I'd also get behind the second option. The ilter icon seems like a good way to mitigate chances for confusion, though I still think some sort of explicit explanation somewhere about what's going on wouldn't hurt.

**Comment by Reptilioo** [comment]
[] Add top button link to search panel
[] Add tooltip explaining the filter bar purpose
[] Move filter bar from top of page to top of sutta list and handle its display
[] Modify `advanced-search.html` as `search-panel.html` ?

**Comment by interested-in-dhamma** [comment]
Renamed the title to be more fitting

**Comment by Reptilioo** [comment]
The design is not set in stone but here is how it could look like: 

![Image](https://github.com/user-attachments/assets/9eb83355-f84a-4e15-a073-bf4906973e86)

I'm currently wondering what I should put in the tooltip since the text in the filter bar is quite explicit, though not giving all the informations. It could explicitly say that the filter is on citation, english title, pali title, and headings?

**Comment by interested-in-dhamma** [comment]
Looks great.

I would simply name it "Search"

Yes, the icons not being filled would be better UI/UX wise.

I am not sure about the tooltip. One thing to reconsider here maybe would be to show everything in the list of suttas that can be filtered on (I think only now pali titles are missing?), and then the filter input can simply say "Filter suttas..."
Then you would need to check how it looks with the pali titles there. Could be something to try it, but perhaps the names get too long.

**Comment by Reptilioo** [comment]
Second try for the icons, also changed "Dark Mode" to "Theme"

![Image](https://github.com/user-attachments/assets/acecb36a-3eca-4808-89c0-d1f4f347f227)

**Comment by interested-in-dhamma** [comment]
Looks good.

It's not really related to this issue, but it makes sense now to color the one that is currently selected.
Maybe use bold text for the one selected.
Or what is often used to emphasize one thing is to de-emphasize others, so in this case the non-selected can be gray and the selected black.

**Comment by bkhpanigha** [comment]
I would say we would need to either add the tooltip to explicitly tell the user that they can search Pali titles _or_ show the Pali titles on the list. But I think the latter option, at least with the way titles are being displayed now, will make things look cluttered.

**Comment by Reptilioo** [comment]
I agree with the clutteredness of displaying pali titles, so I implemented the tooltip method.

![Image](https://github.com/user-attachments/assets/38ab103e-7f0c-498c-be16-0fcbf2bd613f)

Though it could still be improved since the tooltip here is fixed, while we could implement javascript to make it follow the cursor.

**Comment by interested-in-dhamma** [comment]
What's commonly done is adding a label below the input, a "helper text". The problem with tooltips is they don't work when not using a mouse.

**Comment by Reptilioo** [comment]
Yes, maybe I could just add a (i) button like the ones we have next to the top right buttons. With a pop up giving the explanation, so mobile users would see it too.

As for headings I agree.

**Comment by interested-in-dhamma** [comment]
Wouldn't that text actually just fit as placeholder? It doesn't seem to be too long.

**Comment by Reptilioo** [comment]
It does, font-size would just need to be reduced on mobile, which is expected.

![Image](https://github.com/user-attachments/assets/a2e57afc-1fad-429f-b972-310ce3f7fbe4)

**Comment by interested-in-dhamma** [comment]
Both look good to me.

**Comment by bkhpanigha** [comment]
@Reptilioo > Yes, maybe I could just add a (i) button like the ones we have next to the top right buttons. With a pop up giving the explanation, so mobile users would see it too.

That's actually what I had in mind. I thought "tooltip" explicitly meant that.

**Comment by Reptilioo** [comment]
@bkhpanigha 

Well, I'm going a bit outside of the scope of this issue, and I by no means say that this should be the final version but here is how it could look like

![Image](https://github.com/user-attachments/assets/b8b2fdd9-c49e-4f22-9cdc-2debd4d3230d)

I put the filter above the sutta list as hinted previously, and also took the liberty to change the font to Roboto without serif (for the elements on this page **only**, as I think it can look better on an UI)

**Comment by interested-in-dhamma** [comment]
@bkhpanigha @Reptilioo 

Draft 2 looks good to me.

Another presentation to consider is the way ATI does it:

![Image](https://github.com/user-attachments/assets/f256d317-823b-43bb-85a9-5eeae473bc81)

I always liked the simplicity of that. What do you guys think of it?

**Comment by Reptilioo** [comment]
That's a bit too messy and overwhelming for my taste, but this might be just me.

**Comment by bkhpanigha** [comment]
@Reptilioo I would opt for draft two with one modification: having the Sutta number on the upper row of each card, before the English name. The current "headings" could eventually become a short description similar to the ATI page that @interested-in-dhamma posted.

**Comment by Reptilioo** [comment]
Sure, I will make a draft with the sutta number before the english name, on the same line, and put text below it see how it could look.

If we remove headings we could then put the pali name in place of the headings. 
Though, I find the way the headings are currently used to be pretty neat, concise enough to know at first glance what is going to be covered without overloading the UI with text and cards of different sizes.
Also the search panel uses the headings so it would need to be modified, though the changes should be simple.

A proposal for the use of short descriptions: While keeping the headings, or not. We could add a caret pointing down at the bottom of the cards possessing a description, clicking on it would unfold the description part.

**Comment by Reptilioo** [comment]
## Draft 3

![Image](https://github.com/user-attachments/assets/eee15802-e54e-4599-a118-86f8498f32ff)

Removed headings, added description part, modified pali title color

**Comment by interested-in-dhamma** [comment]
That looks good. 
Maybe a draft 4 with notation like: "MN1: Mūlapariyāyasutta — The Explanation of the Root" 
(without the subtitle, to see how it looks.)

I like the separation for the description.

**Comment by Reptilioo** [comment]
## Draft 4

![Image](https://github.com/user-attachments/assets/fe066c94-1054-490f-aa38-8e56467bf0f9)

**Comment by bkhpanigha** [comment]
I'd go for Draft 3-2. The sans serif font clearly can't cope with some Pali diacritics (See newest card in "What's New".

Is it possible to just keep the headings for now, though? I won't have time to write the longer descriptions any time soon, I think.

**Comment by Reptilioo** [comment]
@bkhpanigha Sure, we could even have the heading not showing whenever there is a description on the current sutta.

Also, even if we have descriptions for all the suttas, we could keep the headings only for the search panel, as the search is also being done on them and I believe it can serve as a keyword to find a sutta on a specific theme while the specific word for this theme doesn't appear in the sutta, e.g. "dependent origination" for MN 38, "training" for MN 39.

Additionnally, I could maybe help for the descriptions, between the sutta content, the heading and the comments, I may be able to write a couple of lines without misrepresenting the Buddha's teaching. But you would still need to review them if you happen to have some time in the future. No offense in declining.

**Comment by bkhpanigha** [comment]
@Reptilioo 
> Additionnally, I could maybe help for the descriptions, between the sutta content, the heading and the comments, I may be able to write a couple of lines without misrepresenting the Buddha's teaching. But you would still need to review them if you happen to have some time in the future. No offense in declining.

Sounds good actually, we can do that.

**Comment by uruvelakassapa** [comment]
Potentially a bad idea but it could also be worth putting through the translation and comment files (perhaps rendered together) programmatically to a LLM along with what good output looks like for a given input (the more examples the better [within reason]). Might save some time to have a starting point which can then be further refined.

**Comment by Reptilioo** [comment]
@uruvelakassapa It is a great idea if speed is important, but if it's not, I may prefer to go the old way with this one.
